dev-guide/feature-gate.md

```diff
@@ -0,0 +1,293 @@
+3. allow feature gates to move from TechPreviewNoUpgrade to Default without releases as TechPreviewNoUpgrade first.
+
+### Non-Goals
+
+1. Change thresholds for requiring TechPreviewNoUpgrade.
+   This is a possible future goal, but an update here would be required.
+2. Change thresholds for promoting from TechPreviewNoUpgrade to Default.
+   This is a possible future goal, but an update here would be required.
+
+## Proposal
+### Phase 1
+For developers using this, the first phase will look like
+1. Open PR to openshift/api to add your feature gate to [TechPreviewNoUpgrade](https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L117).
+   The PR should be confined to just the feature gate change and should include a link to a merged enhancement.
+2. Nag api-approvers with link to your PR right away and then every 24h or so until they merge it.
+3. Vendor the change into openshift/cluster-config-operator.
+4. Nag api-approvers with link to your PR right away and then every 24h or so until they merge it.
+
+Notice that this flow eliminates vendoring into multiple repositories and makes it possible to test a feature gate change
+with a single PR, so /payload testing functions properly.
+
+### Phase 2
+1. Open PR to openshift/api to add your feature gate to [TechPreviewNoUpgrade](https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L117).
+   The PR should be confined to just the feature gate change and should include a link to a merged enhancement.
+2. Nag api-approvers with link to your PR right away and then every 24h or so until they merge it.
+3. Automation vendors openshift/api into cluster-config-operator and opens a PR in a few hours.
+   If policy allows it, machine /lgtm-ing is also desired.
+
+Notice that we're down to a single PR to add or promote a featuregate.
+
+### Mechanics of how it works
+The reason that the current state of determining which feature gates are enabled for which feature sets is handled
+by vendoring code as opposed to API status, is that when feature gates are promoted between releases the vN-1 will
+have a pre-GA version.
+So if vN writes status that indicates a feature gate should be on, the vN-1 must not honor that.
+At the time, we had so few feature gates (see the last 12 releases), that this didn't matter, but since we've added
+TechPreviewNoUpgrade CI jobs, stabilized them, and developed means of inspecting reliability of tests and jobs, there
+is now a benefit of using these capabilities to test our code before impacting real clusters.
+
+
+To overcome the original limitation, we will need to key the enabled/disabled feature gates with the payload version
+that they are for.
+This will ensure that older versions do not enable feature gates that were not GA until the later version.
+To do this we will
+1. Update FeatureGateStatus to contain a list of enabled and disabled feature gates for up to every version listed
+   in the CVO history.
```

Comment:
ClusterVersion `status.history` is not exhaustive: openshift/cluster-version-operator#791, openshift/cluster-version-operator#805, #1153.  It's possible that the fraction of history maintained by #1153 is sufficient for this enhancement.  And it's possible that implicit in this enhancement is that `status.featureGates` will be pruned to keep up with ClusterVersion `status.history` pruning.

I'm not all that clear on why you'd want such deep history in `featureGates` though.  Wouldn't the most recent `Completed` version, and any subsequent `Partial`  versions be sufficient?  Or possibly the useful version horizon extends back to the previous minor version, to accommodate possible eventual support for [rollbacks within a z stream](https://issues.redhat.com/browse/OTA-455)?  But once you've `Completed`  a 4.y.z, I don't understand why any 4.(y-1) or older feature gate information would matter.